src/StringTools/OpportunisticIntern.cs

```diff
@@ -0,0 +1,295 @@
+            /// </summary>
+            private readonly bool _gatherStatistics;
+
+            /// <summary>
+            /// Number of times interning with hardcoded string literals worked.
+            /// </summary>
+            private int _hardcodedInternHits;
+
+            /// <summary>
+            /// Number of times the regular interning path found the string in the cache.
+            /// </summary>
+            private int _regularInternHits;
+
+            /// <summary>
+            /// Number of times the regular interning path added the string to the cache.
+            /// </summary>
+            private int _regularInternMisses;
+
+            /// <summary>
+            /// Number of times interning wasn't attempted.
+            /// </summary>
+            private int _rejectedStrings;
+
+            /// <summary>
+            /// Total number of strings eliminated by interning.
+            /// </summary>
+            private int _internEliminatedStrings;
+
+            /// <summary>
+            /// Total number of chars eliminated across all strings.
+            /// </summary>
+            private int _internEliminatedChars;
+
+            /// <summary>
+            /// Maps strings that went though the regular (i.e. not hardcoded) interning path to the number of times they have been
+            /// seen. The higher the number the better the payoff if the string had been hardcoded.
+            /// </summary>
+            private Dictionary<string, int> _missedHardcodedStrings;
+
+#endregion
+
+            public WeakStringCacheInterner(bool gatherStatistics)
+            {
+                if (gatherStatistics)
+                {
+                    _missedHardcodedStrings = new Dictionary<string, int>();
+                }
+                _gatherStatistics = gatherStatistics;
+            }
+
+            /// <summary>
+            /// Intern the given internable.
+            /// </summary>
+            public string InterningToString(ref InternableString candidate)
+            {
+                if (_gatherStatistics)
+                {
+                    return InternWithStatistics(ref candidate);
+                }
+                else
+                {
+                    TryIntern(ref candidate, out string result);
+                    return result;
+                }
+            }
+
+            /// <summary>
+            /// Returns a string with human-readable statistics.
+            /// </summary>
+            public string FormatStatistics()
+            {
+                StringBuilder result = new StringBuilder(1024);
+
+                string title = "Opportunistic Intern";
+
+                if (_gatherStatistics)
+                {
+                    result.AppendLine(string.Format("\n{0}{1}{0}", new string('=', 41 - (title.Length / 2)), title));
+                    result.AppendLine(string.Format("||{0,50}|{1,20:N0}|{2,8}|", "Hardcoded Hits", _hardcodedInternHits, "hits"));
```

Comment:
Should all these strings be localized?

Reply:
For debug output behind an undocumented environment variable I vote for no. There are many places in the codebase producing non-localized debug output, for example `DumpSchedulerState`.